.github/workflows/macstadium-android-e2e.yml

```diff
@@ -1,56 +1,82 @@
+      - name: Checkout repo
+        uses: actions/checkout@v4
 
       - name: Set up github keys
         run: git config core.sshCommand "ssh -i ~/.ssh/id_ed25519 -F /dev/null"
 
+      - name: Clean Android app
+        run: yarn clean:android > /dev/null 2>&1 || true
+
       - name: Set up ENV vars & scripts
+        env:
+          CI_SCRIPTS_RN_UPGRADE: ${{ secrets.CI_SCRIPTS_RN_UPGRADE }}
         run: |
-          # read local env vars
-          source ~/.bashrc
-          # fetch env vars
+          source ~/.zshrc
           git clone git@github.com:rainbow-me/rainbow-env.git
-          # unpack dotenv
-          mv rainbow-env/android/app/google-services.json android/app
           mv rainbow-env/dotenv .env && rm -rf rainbow-env
-          # run CI scripts
-          eval $CI_SCRIPTS
-          # tweak dotenv for e2e
-          sed -i''-e "s/\IS_TESTING=false/IS_TESTING=true/" .env && rm -f .env-e
-          # set up password
-          cp android/keystores/debug.keystore android/keystores/rainbow-key.keystore
-          sed -i -e "s:rainbow-alias:androiddebugkey:g" android/app/build.gradle
-          export RAINBOW_KEY_ANDROID_PASSWORD=android
-      - name: Install deps via Yarn
-        run: yarn setup-ci
+          eval $CI_SCRIPTS_RN_UPGRADE
+          sed -i'' -e "s/IS_TESTING=false/IS_TESTING=true/" .env && rm -f .env-e 
+      
+      - name: Get Yarn cache directory path
+        id: yarn-cache-dir-path
+        run: echo "dir=$(yarn config get cacheFolder)" >> $GITHUB_OUTPUT
+
+      - name: Cache Yarn dependencies
+        uses: actions/cache@v4
+        with:
+          path: |
+            ${{ steps.yarn-cache-dir-path.outputs.dir }}
+            .yarn/cache
+            .yarn/install-state.gz
+          key: ${{ runner.os }}-yarn-${{ hashFiles('**/yarn.lock') }}
+          restore-keys: |
+            ${{ runner.os }}-yarn-
+
+      - name: Install dependencies
+        run: |
+          yarn cache clean --all && yarn install && yarn setup
+
+      - name: Check for frozen lockfile
+        run: ./scripts/check-lockfile.sh
+
+      - name: Audit CI
+        run: yarn audit-ci --config audit-ci.jsonc
+
+      - name: Lint
+        run: yarn lint:ci
+
+      - name: Unit tests
+        run: yarn test
 
       - name: Rebuild detox cache
         run: ./node_modules/.bin/detox clean-framework-cache && ./node_modules/.bin/detox build-framework-cache
+      
+      - name: Version debug
+        run: |
+          npx react-native info
 
-      - name: Build the app in Release mode
-        run: ./node_modules/.bin/detox build --configuration android.emu.release
+      - name: Fix permissions
+        run: |
+          chmod -R +x node_modules/react-native/scripts
+          chmod -R +x node_modules/@sentry/react-native/scripts
 
-      - name: Run Android e2e tests
-        run: ./node_modules/.bin/detox test -R 5 --configuration android.emu.release --forceExit
+      - name: Build the app in Release mode
```

Comment:
For now this is ok but we should split everything else from here and above into an action so we can create two separate jobs (android / iOS) that happen after (we install all deps only once) then separate jobs to build each platforms.